mobile/library/common/http/client.cc

```diff
@@ -437,16 +439,27 @@ void Client::DirectStreamCallbacks::latchError() {
   }
   const auto& info = request_decoder->streamInfo();
 
+  std::vector<std::string> error_msg_details;
   if (info.responseCode().has_value()) {
     error_->error_code = Bridge::Utility::errorCodeFromLocalStatus(
         static_cast<Http::Code>(info.responseCode().value()));
+    error_msg_details.push_back(absl::StrCat(info.responseCode().value()));
   } else if (StreamInfo::isStreamIdleTimeout(info)) {
     error_->error_code = ENVOY_REQUEST_TIMEOUT;
   } else {
     error_->error_code = ENVOY_STREAM_RESET;
   }
 
-  error_->message = info.responseCodeDetails().value_or("");
+  error_msg_details.push_back(absl::StrCat(error_->error_code));
+  if (info.responseCodeDetails().has_value()) {
+    error_msg_details.push_back(info.responseCodeDetails().value());
+  }
+  // The format of the error message propogated to callbacks is:
+  //  {RESPONSE_CODE}|{ERROR_CODE}|{DETAILS}
+  // Where RESPONSE_CODE is the HTTP response code from StreamInfo::responseCode().
+  // ERROR_CODE is of the envoy_error_code_t enum type, and gets mapped from RESPONSE_CODE.
+  // DETAILS is the contents of StreamInfo::responseCodeDetails().
+  error_->message = absl::StrJoin(std::move(error_msg_details), "|");
```

Comment:
Instead of doing string concatenation, can we simply update the struct to include the response code, error code, and details?

Reply:
We could, but this `message` field basically populates the free form "error details" that we use for constructing the detailed error message (https://github.com/envoyproxy/envoy/pull/33871/).  I wanted to make it easier to have the same error details on iOS and Android, so that's why I'm formatting it in this free form details message field that gets passed around in the EnvoyError object.

Reply:
I still think adding some members in the struct is better, but given that we may refactor this in the future, I agree that it's a bit too premature to optimize this, so I'm fine with the free-form format for now.

What about having a format like `response_code=<...>|error_code=<..>|details=<...>` or something like that?It's easy to get mixed up between response code and error code and having to know which position it is a bit unintuitive IMO, but I'll leave it up to you to decide :)

Reply:
I was debating in my mind whether is should add labels to the fields.  On the one hand, it makes it clearer, on the other hand, I didn't want wasted text that didn't convey meaning.  But this discussion made me think having the labels is better, it reads nicer for someone not familiar, so I made the change.  Thanks for the suggestion!